File: leetcode_ly/c++/Bit/SingleNumberThree.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int result = 0;
        for(int i = 0; i < nums.size(); i++){
            result ^= nums[i];
        }
        int pos = result & ~(result - 1);
        vector<int> ret(2,0);
        for(int j = 0; j < nums.size(); j++){
            if(pos & nums[j]){
                ret[0] ^= nums[j];
            }
            else{
                ret[1] ^= nums[j];
            }
        }
        return ret;
    }
};

File: leetcode_ly/java/Bit Manipulation && Design/SingleNumberThree.java
public class Solution {
    public int[] singleNumber(int[] nums) {
        int[] res = new int[2];
        if(nums == null || nums.length < 2) return res;
        int val = 0;
        for(int num : nums){
            val ^= num;
        }
        int index = 0;
        while(val != 0){
            if((val & 1) == 1) break;
            else{
                val = val >>> 1;
                index++;
            }
        }
        int num1 = 0;
        int num2 = 0;
        for(int num : nums){
            if(((num >>> index) & 1) == 1){
                num1 ^= num;
            }
            else{
                num2 ^= num;
            }
        }
        res[0] = num1;
        res[1] = num2;
        return res;
    }
}